src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -748,67 +747,73 @@ class RandomStringFromDataset : public Generator<std::string> {
     RandomStringFromDataset(const Node& node, GeneratorArgs generatorArgs)
         : _rng{generatorArgs.rng},
           _id{generatorArgs.actorId},
-          _path{node["path"].maybe<std::string>().value()},
-          _pathLength{_path.size()}{}
+          _path{node["path"].maybe<std::string>().value()} {
+        loadDataset();
+    }
 
+private:
     void loadDataset() {
-      {
-        if (_pathLength <= 0) {
-            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
-            "ChooseFromDataset requieres non-empty path"));
+        {
+            if (_path.empty()) {
+                BOOST_THROW_EXCEPTION(
+                    InvalidValueGeneratorSyntax("ChooseFromDataset requieres non-empty path"));
             }
 
-        {
-            if (_all_datasets.count(_path) == 0){
-                _ifs.open(_path, std::ifstream::in);
-                if (_ifs.is_open()){
-                    while(std::getline(_ifs, _current_line))
-                    {
-                        // Ignore emtpy lines
-                        if (_current_line.size()!=0){
-                            _all_datasets[_path].push_back(_current_line);
+            /* If path the doesn't exist in memory we load it.
+               To avoid reading and storing the datasets multiple times
+               we check if the path already exists in the map (as key). If
+               it doesn't exist, we open the file and store its contents.
+            */
+
+            {
+                std::mutex _dataset_mutex;
+                std::lock_guard<std::mutex> lk(_dataset_mutex);
+
+                if (_all_datasets.count(_path) == 0) {
```

Comment:
1. I will add it. I had it before, but lost if after moving the code around.
2. I don't think this will work. If I add an empty vector, some of the threads will continue to evaluate() as if the data was already there and the data could still be not loaded, or only partially, by the first thread. The results could be difficult to predict. And we won't have repeatable outcomes on different runs.

Reply:
There won't be any calls to evaluate until after all the actors and phases and everything are constructed. There's a hard barrier in there. So, no risk of a call to evaluate with an empty vector.

Reply:
Oh, didn't know that. Then, I will implement that too. Thanks.

Reply:
My understanding (could be wrong) ...

The first time you call `_all_datasets[_path].push_back(currentLine);`  for a given `_path` a new `std::vector<std::string>` is implicitly created an emplaced in the vector `_all_datasets` at key `_path`. I'm not suggesting calling `_all_datasets[_path].push_back(std::vector<std::string>());` but rather doing something like `_all_datasets[_path] = std::vector<std::string>()`. The end result should be _all_datasets[_path] temporarily having an empty vector, and no difference once you start emplacing things. Does that make sense? Hopefully it's even correct. 

Honestly, the implicit creation surprised me a bit, as I probably would have made a local `auto local_vector = std::vector<std::string>`, loaded in all the values, and then done `_all_datasets[_path] = std::move(local_vector)` or some such.

Reply:
Mmm well, I think that I can just remove the first element of the vector after the load has finished, and then there won't be problems.

Reply:
Ok, so, what I am going to do is add an empty vector. Then in the last line of the method, once all the data is loaded, I remove the empty vector that is at the first position.

Reply:
Looks beautiful. Thanks.